pkg/util/clusterdata/ingress_profiles_task.go

```diff
@@ -0,0 +1,118 @@
+	"context"
+
+	operatorv1 "github.com/openshift/api/operator/v1"
+	operatorclient "github.com/openshift/client-go/operator/clientset/versioned"
+	"github.com/sirupsen/logrus"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+
+	"github.com/Azure/ARO-RP/pkg/api"
+)
+
+func newIngressProfileEnricherTask(log *logrus.Entry, restConfig *rest.Config, oc *api.OpenShiftCluster) (enricherTask, error) {
+	operatorcli, err := operatorclient.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	kubecli, err := kubernetes.NewForConfig(restConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	return &ingressProfileEnricherTask{
+		log:         log,
+		operatorcli: operatorcli,
+		kubecli:     kubecli,
+		oc:          oc,
+	}, nil
+}
+
+type ingressProfileEnricherTask struct {
+	log         *logrus.Entry
+	operatorcli operatorclient.Interface
+	kubecli     kubernetes.Interface
+	oc          *api.OpenShiftCluster
+}
+
+func (ef *ingressProfileEnricherTask) FetchData(ctx context.Context, callbacks chan<- func(), errs chan<- error) {
+	// List IngressControllers from  openshift-ingress-operator namespace
+	// Each IngressController will be the basis for an IngressProfile with the below mapping:
+	//     IngressController.Name -> IngressProfile.Name
+	//     IngressController.spec.endpointPublishingStrategy.loadBalancer.scope -> IngressProfile.Visibility
+	//             Internal -> Private
+	//             External -> Public
+	ingressControllers, err := ef.operatorcli.OperatorV1().IngressControllers("openshift-ingress-operator").List(ctx, metav1.ListOptions{})
+	if err != nil {
+		ef.log.Error(err)
+		errs <- err
+		return
+	}
+
+	// List Services from openshift-ingress namespace
+	// Among those Services, look for the ones of type LoadBalancer, with label "app: router". The matching will be done with
+	// IngressController based on ingresscontroller.operator.openshift.io/owning-ingresscontroller label.
+	// Service IP will be taken from the candidate and added to the corresponding IngressProfile
+	services, err := ef.kubecli.CoreV1().Services("openshift-ingress").List(ctx, metav1.ListOptions{
+		LabelSelector: "app=router",
+	})
+
+	if err != nil {
+		ef.log.Error(err)
+		errs <- err
+		return
+	}
+
+	routerIPs := make(map[string]string)
+	for _, service := range services.Items {
+		if len(service.Status.LoadBalancer.Ingress) == 0 {
+			continue
+		}
+		matchingICName := service.ObjectMeta.Labels["ingresscontroller.operator.openshift.io/owning-ingresscontroller"]
+		routerIPs[matchingICName] = service.Status.LoadBalancer.Ingress[0].IP
+	}
+
+	// Reconcile IngressController and corresponding router service
+	ingressProfiles := make([]api.IngressProfile, len(ingressControllers.Items))
+	for i, ingressController := range ingressControllers.Items {
+		ingressProfiles[i] = api.IngressProfile{
+			Name: ingressController.ObjectMeta.Name,
+			IP:   routerIPs[ingressController.ObjectMeta.Name],
```

Comment:
There is theoretical possibility this will not be present in the map. 
If Ingress controller was deleted, but for some reason service deletion is stuck due to bad finalizee or azure being slow (deleting old IP address). Usually this will not happen if user didn't intervene and delete other object.

I think check here woud be good too.

Reply:
Hi @mjudeikis , yes, this is in theory possible and in the first version of the PR, I even checked the presence of the IP in the map and logged a message in case it was not there. However, @jim-minter suggested that this case so unusual that it did not deserve a log and proposed to assign the IP the way I do it now, right away.